File: backend/models/decisions.py
"""Decision-log models — read-only audit view over recommendations + learning store.

Field names and enum values mirror `frontend/src/types/decisions.ts` so the
decision log and pre-approval packet screens can bind directly.
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field


class DecisionStatus(str, Enum):
    pending = "pending"
    approved = "approved"
    rejected = "rejected"
    expired = "expired"


class StrategyModule(str, Enum):
    """Frontend-facing module ids (`stat_arb` is reserved for a later phase)."""

    stat_arb = "stat_arb"
    volatility = "volatility"
    gamma_scalping = "gamma_scalping"
    vega_scalping = "vega_scalping"


class GateStatus(str, Enum):
    pass_ = "pass"
    fail = "fail"
    warn = "warn"


class RagCitation(BaseModel):
    document: str
    chapter: str | None = None
    page: int | None = None


class OptionLeg(BaseModel):
    leg_id: int
    type: str
    position: int
    strike: float | None = None
    expiration: str | None = None
    delta: float
    gamma: float
    theta: float
    vega: float
    mid_price: float | None = None


class GreeksTotals(BaseModel):
    total_delta: float = 0.0
    total_gamma: float = 0.0
    total_theta: float = 0.0
    total_vega: float = 0.0


class TradeEconomics(BaseModel):
    net_edge_after_costs: float
    margin_estimate: float
    expected_gamma_pnl: float = 0.0
    expected_theta_loss: float = 0.0
    total_transaction_cost: float = 0.0
    slippage_cap_bps: float = 0.0


class TradePlan(BaseModel):
    stop: str
    target: str
    time_exit: str


class RiskGate(BaseModel):
    id: str
    label: str
    status: GateStatus
    detail: str | None = None


class StrategyContext(BaseModel):
    iv_annualized: float | None = None
    garch_forecast: float | None = None
    gamma_theta_breakeven_pct: float | None = None
    entry_mode: str | None = None
    term_structure_ok: bool | None = None
    vega_neutral_at_entry: bool | None = None
    intraday_iv_z_score: float | None = None
    must_flatten_same_day: bool | None = None


class DecisionRecord(BaseModel):
    """One audited decision — matches the frontend `PendingDecision` shape."""

    decision_id: str
    signal_id: str
    status: DecisionStatus
    module: StrategyModule
    underlying_symbol: str
    confidence: float
    regime: str
    explanation: str
    rationale: str
    rag_citations: list[RagCitation] = Field(default_factory=list)
    legs: list[OptionLeg] = Field(default_factory=list)
    totals: GreeksTotals = Field(default_factory=GreeksTotals)
    economics: TradeEconomics
    plan: TradePlan
    event_risks: list[str] = Field(default_factory=list)
    failure_modes: list[str] = Field(default_factory=list)
    strategy_context: StrategyContext = Field(default_factory=StrategyContext)
    risk_gates: list[RiskGate] = Field(default_factory=list)
    created_at: datetime
    expires_at: datetime
    scenario_tag: str | None = None
    alternative_strategies: list[str] | None = None

File: frontend/src/types/decisions.ts
export type StrategyModule =
  | "stat_arb"
  | "volatility"
  | "gamma_scalping"
  | "vega_scalping";

export type EntryMode =
  | "standard"
  | "cheap_vol_mode"
  | "earnings_gap_mode"
  | "high_realized_vol_mode";

export type DecisionStatus = "pending" | "approved" | "rejected" | "expired";

export type GateStatus = "pass" | "fail" | "warn";

export interface RagCitation {
  document: string;
  chapter?: string;
  page?: number;
}

export interface OptionLeg {
  leg_id: number;
  type: "call" | "put" | "stock";
  position: number;
  strike?: number;
  expiration?: string;
  delta: number;
  gamma: number;
  theta: number;
  vega: number;
  mid_price?: number;
}

export interface GreeksTotals {
  total_delta: number;
  total_gamma: number;
  total_theta: number;
  total_vega: number;
}

export interface TradeEconomics {
  net_edge_after_costs: number;
  margin_estimate: number;
  expected_gamma_pnl: number;
  expected_theta_loss: number;
  total_transaction_cost: number;
  slippage_cap_bps: number;
}

export interface TradePlan {
  stop: string;
  target: string;
  time_exit: string;
}

export interface RiskGate {
  id: string;
  label: string;
  status: GateStatus;
  detail?: string;
}

export interface StrategyContext {
  /** Stat arb */
  residual_z_score?: number;
  hedge_ratio?: number;
  half_life_days?: number;
  cointegrated_windows?: number;
  economics_aligned?: boolean;
  pair_y?: string;
  pair_x?: string;
  /** Volatility */
  iv_annualized?: number;
  garch_forecast?: number;
  gamma_theta_breakeven_pct?: number;
  /** Gamma */
  entry_mode?: EntryMode;
  term_structure_ok?: boolean;
  vega_neutral_at_entry?: boolean;
  /** Vega */
  intraday_iv_z_score?: number;
  must_flatten_same_day?: boolean;
}

export interface PendingDecision {
  decision_id: string;
  signal_id: string;
  status: DecisionStatus;
  module: StrategyModule;
  underlying_symbol: string;
  confidence: number;
  regime: string;
  explanation: string;
  rationale: string;
  rag_citations: RagCitation[];
  legs: OptionLeg[];
  totals: GreeksTotals;
  economics: TradeEconomics;
  plan: TradePlan;
  event_risks: string[];
  failure_modes: string[];
  strategy_context: StrategyContext;
  risk_gates: RiskGate[];
  created_at: string;
  expires_at: string;
  scenario_tag?: string;
  alternative_strategies?: string[];
}

export interface BotStatus {
  execution_mode: "shadow" | "paper" | "live";
  supervision_mode?: "supervised" | "semi_autonomous" | "fully_autonomous";
  autonomy: "full" | "supervised";
  scheduler_mode: "active" | "learning" | "reduced" | "paused";
  regime: string;
  daily_pnl: number;
  win_rate: number;
  drawdown_pct: number;
  portfolio_greeks: GreeksTotals;
  circuit_breakers_active: string[];
  pending_count: number;
  one_trade_locked: boolean;
  active_trade_id?: string | null;
  kill_switch_armed?: boolean;
  api_health?: string;
}

export const STRATEGY_LABELS: Record<StrategyModule, string> = {
  stat_arb: "Statistical Arbitrage",
  volatility: "Simple Volatility",
  gamma_scalping: "Gamma Scalping",
  vega_scalping: "Vega Scalping",
};

export const ENTRY_MODE_LABELS: Record<EntryMode, string> = {
  standard: "Standard",
  cheap_vol_mode: "Cheap Vol",
  earnings_gap_mode: "Earnings Gap",
  high_realized_vol_mode: "High Realized Vol",
};
